labs/lab9/9.md

```diff
@@ -0,0 +1,250 @@
+
+You can change ports on this development server by editing basicapp.py file. In the line "app.run()", try adding argument port=5001, or debug=True. Control-C to exit.
+
+To get our uWSGI running, let's install it through apt.
+
+    apt install uwsgi uwsgi-plugin-python
+
+Next, we need to edit basicapp.ini, your uWSGI config file. Then we'll move it to a location accessible to the uwsgi server. 
+
+Check uwsgi.py exist in your directory. This file contains your callable function app. This is called by uWSGI to reach your app. 
+
+Open 'basicapp.ini' file. This is the config file for uWSGI. Look for 'wsgi-file' variable. This is set to your uwsgi.py file. The 'callable' variable should be set to 'app'. Equivalently, 'module = basicapp:app' should give you the same result. 
+
+The 'chdir' variable should be set to your project directory. uWSGI startup file will change the working directory to this assigned location. Any other relative file path will be relative to this location. 
+
+If you have project name different than 'basicapp', change every instance of 'basicapp' to your app name. 
+
+'pythonpath' and 'virtualenv' helps uWSGI to find the right packages.
+
+In Debian based systems, Apache2 and Nginx are assigned 'www-data' for their user and group. In our uwsgi config, we set 'gid' and 'uid' to 'www-data'. That way when systemd runs the startup script, it would know what user and group to set uWSGI process to.  
+
+'master = true' respawns processes if the processes die. uWSGI starts a master processes that spawn application server processes, and it monitors these children processes in case of failure.
+
+'processes' and 'threads' tells uWSGI master the number of processes to spawn and how many threads per process. It's arbitrary set at this point. It should increase as your application demand increase, up to your system resource limit. 
+
+'socket' is where uWSGI and Nginx communicates. It can be Unix sockets or internet sockets. As long as nginx is set to the same location. Right now, 0.0.0.0:8370 is assigned. The IP 0.0.0.0 tells the uWSGI server to attach to no specific IP. That means it's accessible from anywhere in the network. It can be used to communicate reverse proxy located on a different physical server.
+
+For testing purposes, lets comment out the 'socket' line. And add a line for http-socket. 
+
+    http-socket = :8370
+
+Save this file and we run uWSGI with this configuration. First, check if a uwsgi process is already running. If so kill it. Either by `systemctl stop uwsgi` or `kill -9` 
+
+    uwsgi --ini basicapp.ini
+
+You should be able to see your app running on port 8370. (ports I choose is arbitrary, Just as long as it's not occupied by another process or below 1024) 
+
+If you do not see your flask app, it might be because we have not changed our project folder owner and groups yet. Change the folder group to 'www-data'. And chmod with a group read, write, and execute permission. Rerun uwsgi after the changes.
```

Comment:
"chown the folder to the `www-data` group, and give that group read, write and execute permissions"